### ISSUE TITLE
Retrieving Objects from Claims fields

### ISSUE THREAD

**Issue by elithrar** [dgrijalva/jwt-go#64]
Is it possible to marshal an object from a `t.Claims` value (which is a `map[string]interface{}`) back to its original type? 

Typically this would just be a type assertion away - e.g.

```go

type User struct {
    ID string
    Data string
    Admin bool
}

// Generation
user = &User{ID: "1313193933391913", Data: "matt@example.com", Admin: false}
t.Claims["user"] = user

// Parsing after calling jwt.Parse(...)
val, ok := t.Claims["user"].(*User)
if !ok {
    return nil, ErrInvalidUser
}

// val is of type *User.
return val, nil
```

My assumption here is that the boolean field in the struct is being marshalled to a string and cannot be un-marshalled back to the struct based on the below:

```
fmt.Printf("Type: %t\nValue: %v\n", t.Claims["user"], t.Claims["user"])

// Outputs:
Type: map[%!t(string=ID):%!t(string=1313193933391913) %!t(string=Data):%!t(string=matt@example.com) %!t(string=Admin):false]
Value: map[ID:1313193933391913 Data:matt@example.com Admin:false]
```

Is there a way to resolve this without writing a ton of un-marshalling logic?

**Comment by dgrijalva** [comment]
Unfortunately, not at this time.  This has come up a few times before.  See #26 for details.  I've run into this in a few places, including the upcoming support for JWK.  I'm all ears if you have a good idea.

**Comment by m-mmmmmm** [comment]
Add the following methods:
````
    func (t *Token) RawClaim() []byte {}
    func (t *Token) RawHeader() []byte {}
````
Populate the value during Parse() with [claimBytes](https://github.com/dgrijalva/jwt-go/blob/master/jwt.go#L104) and [headerBytes](https://github.com/dgrijalva/jwt-go/blob/master/jwt.go#L95). Then people can do their own json.Unmarshal outside the library.

**Comment by elithrar** [comment]
Not a terrible idea at all. Since `claimbytes` and `headerbytes` already
exist there's minimal overhead in providing those functions (outside of a
larger API, that is). Would certainly make it easier to extract complex
types from Claims.

**Comment by dgrijalva** [comment]
You can just make them properties. Since the raw value is a string, they can just be substrings and won't cost much. 

-dave

**Comment by crxpandion** [comment]
What about making Token.Claims an interface? That way we can just leverage the stdlib json marshaling logic to detect types. I'll submit a PR to make this more clear.

**Comment by dgrijalva** [comment]
Unfortunately, several of the claims (`exp`, `nbf`, etc) can affect validation.  The validator needs to be able to access them.  Here are some options; none of them are great:

* use `map[string]interface{}`: this is what we do currently.
* parse the claims body twice: we could do this, but it's more expensive.  might make sense to just make the raw claims body accessible.  then the user can opt to do this themselves
* use `interface{}` and try a combination of reflect and type checking to find the necessary keys: this could work as long as the type passed in by the user both contains the necessary fields and is introspectable

I think the first two options are more practical.

**Comment by dgrijalva** [comment]
Reading your PR...  This is a really good idea.  Let me think on it a bit.

**Comment by dgrijalva** [comment]
This has landed on the 3.0.0 branch.  #77
